### PR TITLE
[Merged by Bors] - collect bandwidth metrics per transport

### PR DESCRIPTION
**Pull request by jxs**
## Issue Addressed

Following the conversation on https://github.com/libp2p/rust-libp2p/pull/3666 the changes introduced in this PR will allow us to give more insights if the bandwidth limitations happen at the transport level, namely if quic helps vs yamux and it's [window size limitation](https://github.com/libp2p/rust-yamux/issues/162) or if the bottleneck is at the gossipsub level.
## Proposed Changes

introduce new quic and tcp bandwidth metric gauges.

cc @mxinden (turned out to be easier, Thomas gave me a hint)

### CODE REVIEW

File: beacon_node/network/src/metrics.rs
Comment (mxinden) [inline beacon_node/network/src/metrics.rs]:
```suggestion
    pub static ref INBOUND_LIBP2P_QUIC_BYTES: Result<IntGauge> =
        try_create_int_gauge("libp2p_inbound_bytes", "The inbound bandwidth over libp2p");
```

Best practice would be to use a metric family for all bandwidth, differentiated by a Prometheus label, e.g. `transport="tcp|quic"`.

File: beacon_node/network/src/metrics.rs
Comment (mxinden) [inline beacon_node/network/src/metrics.rs]: This metrics only ever goes up, right? Shouldn't it be a counter instead of a gauge?

**Review by mxinden — COMMENTED** [review]
A bit of a hack, but neat!

Long term, I would consider a proper solution in rust-libp2p the cleaner solution, especially in case you ever introduce a third transport.

That said, for now, ...

File: beacon_node/network/src/metrics.rs
Comment (AgeManning) [reply beacon_node/network/src/metrics.rs]: I think with counters we have to increase the counter, with gauges we can set the value. I think @jxs has to set the value based on the bandwidth at regular intervals. I'll let him comment tho. 

Agree that a counter would be better if we can set it easily.

**Review by AgeManning — COMMENTED** [review]
Nice, this is very helpful. 

Agree with Max's comments

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
pub struct AggregatedBandwidthSinks {
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
impl AggregatedBandwidthSinks {
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
        AggregatedBandwidthSinks {
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Create a new `AggregatedBandwidthSinks`.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total QUIC inbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total TCP inbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total QUIC outbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total TCP outbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total aggregated inbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total aggregated outbound bandwidth.
```

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/metrics.rs]:
```suggestion
    /// Total bandwidth.
```

File: beacon_node/lighthouse_network/src/service/mod.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/service/mod.rs]:
```suggestion
use crate::metrics::AggregatedBandwidthSinks;
```

File: beacon_node/lighthouse_network/src/service/mod.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/service/mod.rs]:
```suggestion
    pub bandwidth: AggregatedBandwidthSinks,
```

File: beacon_node/lighthouse_network/src/service/utils.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/service/utils.rs]:
```suggestion
use crate::metrics::AggregatedBandwidthSinks;
```

File: beacon_node/lighthouse_network/src/service/utils.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/service/utils.rs]:
```suggestion
) -> std::io::Result<(BoxedTransport, AggregatedBandwidthSinks)> {
```

File: beacon_node/lighthouse_network/src/service/utils.rs
Comment (jmcph4) [inline beacon_node/lighthouse_network/src/service/utils.rs]:
```suggestion
        (tcp, AggregatedBandwidthSinks::new(tcp_bandwidth, None))
```

File: beacon_node/network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/network/src/metrics.rs]:
```suggestion
    metrics::AggregatedBandwidthSinks, peer_manager::peerdb::client::ClientKind, types::GossipKind,
```

File: beacon_node/network/src/metrics.rs
Comment (jmcph4) [inline beacon_node/network/src/metrics.rs]:
```suggestion
pub fn update_bandwidth_metrics(bandwidth: &AggregatedBandwidthSinks) {
```

**Review by jmcph4 — CHANGES_REQUESTED** [review]
nit: typos

File: beacon_node/network/src/metrics.rs
Comment (mxinden) [reply beacon_node/network/src/metrics.rs]: You may use [`Counter::inner`](https://docs.rs/prometheus-client/latest/prometheus_client/metrics/counter/struct.Counter.html#method.inner). Note that using a _Gauge_ in a place where a _Counter_ is correct might result in problems e.g. around Grafana's query suggestions.

Comment (jxs) [reply beacon_node/network/src/metrics.rs]: yup it was because of @AgeManning's arguments ehe, managed to switch to a `Counter` and `reset` and `inc_by_value` thanks Max, ptal

File: beacon_node/network/src/metrics.rs
Comment (jxs) [reply beacon_node/network/src/metrics.rs]: yeah makes sense thanks, went ahead and also put the direction into the same metric

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (jxs) [reply beacon_node/lighthouse_network/src/metrics.rs]: thanks @jmcph4 for all these suggestions!

File: beacon_node/lighthouse_network/src/metrics.rs
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/metrics.rs]: just a remaining one

Comment (jxs) [reply beacon_node/lighthouse_network/src/metrics.rs]: thanks! Adressed

**Review by mxinden — COMMENTED** [review]
Cool. Again, long term I think a solution in rust-libp2p would be much cleaner. That said, this should provide us with the numbers we need to evaluate QUIC. :+1:

**Review by AgeManning — APPROVED** [review]
Ok, looks good to me. Lets go

**Comment by AgeManning** [comment]
Just waiting for imminent deneb merge

**Comment by jimmygchen** [comment]
let's go!
bors r+

**Pull request closed without merging**
